src/algebra/module/localized_module.lean

```diff
@@ -0,0 +1,358 @@
+Authors: Jujian Zhang
+-/
+
+import group_theory.monoid_localization
+import ring_theory.localization.basic
+
+/-!
+# Localized Module
+
+Given a commutative ring `R`, a multiplicative subset `S ⊆ R` and an `R`-module `M`, we can localize
+`M` by `S`. This gives us a `localization S`-module.
+
+## Main definitions
+
+* `localized_module.r` : the equivalence relation defining this localization, namely
+  `(m, s) ≈ (m', s')` if and only if if there is some `u : S` such that `u • s' • m = u • s • m'`.
+* `localized_module M S` : the localized module by `S`.
+* `localized_module.mk`  : the canonical map sending `(m, s) : M × S ↦ m/s : localized_module M S`
+* `localized_module.lift_on` : any well defined function `f : M × S → α` respecting `r` descents to
+  a function `localized_module M S → α`
+* `localized_module.lift_on₂` : any well defined function `f : M × S → M × S → α` respecting `r`
+  descents to a function `localized_module M S → localized_module M S`
+* `localized_module.mk_add_mk` : in the localized module
+  `mk m s + mk m' s' = mk (s' • m + s • m') (s * s')`
+* `localized_module.mk_smul_mk` : in the localized module, for any `r : R`, `s t : S`, `m : M`,
+  we have `mk r s • mk m t = mk (r • m) (s * t)` where `mk r s : localization S` is localized ring
+  by `S`.
+* `localized_module.is_module` : `localized_module M S` is a `localization S`-module.
+
+-/
+
+
+namespace localized_module
+
+universes u v
+
+variables {R : Type u} [comm_semiring R] (M : Type v) [add_comm_monoid M] [module R M]
+variables (S : submonoid R)
+
+example (r : R) (m : M) : M := r • m
+
+/--The equivalence relation on `M × S` where `(m1, s1) ≈ (m2, s2)` if and only if
+for some (u : S), u * (s2 • m1 - s1 • m2) = 0-/
+def r (p1 p2 : M × S) : Prop :=
+match p1, p2 with
+| ⟨m1, s1⟩, ⟨m2, s2⟩ := ∃ (u : S), u • s1 • m2 = u • s2 • m1
+end
+
+lemma r.is_equiv : is_equiv _ (r M S) :=
+{ refl := λ ⟨m, s⟩, ⟨1, by rw [one_smul]⟩,
+  trans := λ ⟨m1, s1⟩ ⟨m2, s2⟩ ⟨m3, s3⟩ ⟨u1, hu1⟩ ⟨u2, hu2⟩, begin
+    use u1 * u2 * s2,
+    rw calc (u1 * u2 * s2) • s1 • m3
+        = (u1 * u2 * s2 * s1) • m3 : by simp only [mul_smul]
+    ... = (u1 * s1 * (u2 * s2)) • m3 : by congr' 1; ext; simp only [submonoid.coe_mul]; ring
+    ... = (u1 * s1) • u2 • s2 • m3 : by simp only [mul_smul]
+    ... = (u1 * s1) • u2 • s3 • m2 : by rw [hu2]
+    ... = (u1 * s1 * (u2 * s3)) • m2 : by simp only [mul_smul]
+    ... = (u2 * s3 * (u1 * s1)) • m2 : by congr' 1; ext; simp only [submonoid.coe_mul]; ring
+    ... = (u2 * s3) • u1 • s1 • m2 : by simp only [mul_smul]
+    ... = (u2 * s3) • u1 • s2 • m1 : by rw [hu1]
+    ... = ((u2 * s3) * u1 * s2) • m1 : by simp only [mul_smul],
+    rw [← mul_smul],
+    congr' 1,
+    ext,
+    simp only [submonoid.coe_mul],
+    ring
+  end,
+  symm := λ ⟨m1, s1⟩ ⟨m2, s2⟩ ⟨u, hu⟩, ⟨u, hu.symm⟩ }
+
+
+instance r.setoid : setoid (M × S) :=
+{ r := r M S,
+  iseqv := ⟨(r.is_equiv M S).refl, (r.is_equiv M S).symm, (r.is_equiv M S).trans⟩ }
```

Comment:
Does something like this work? There ought to be some API here...
```suggestion
  iseqv := (r.is_equiv M S).equivalence
```

Reply:
I don't think this works. Maybe I am missing some imports